pkg/irreparable/irreparable.go

```diff
@@ -0,0 +1,150 @@
+	"google.golang.org/grpc/status"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/migrate"
+	dbx "storj.io/storj/pkg/irreparable/dbx"
+)
+
+var (
+	mon = monkit.Package()
+)
+
+// Database implements the irreparable RPC service
+type Database struct {
+	mu     sync.Mutex
+	DB     *dbx.DB
+	logger *zap.Logger
+}
+
+// RemoteSegmentInfo is info about a single entry stored in the irreparable db
+type RemoteSegmentInfo struct {
+	SegmentPath        []byte
+	SegmentDetail      []byte
+	LostPiecesCount    int64
+	RepairUnixSec      int64
+	RepairAttemptCount int64
+}
+
+// NewServer creates instance of Server
+func NewServer(source string, logger *zap.Logger) (*Database, error) {
+	db, err := dbx.Open("sqlite3", source)
+	if err != nil {
+		return nil, err
+	}
+
+	err = migrate.Create("irreparable", db)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Database{
+		DB:     db,
+		logger: logger,
+	}, nil
+}
+
+func (db *Database) locked() func() {
+	db.mu.Lock()
+	return db.mu.Unlock
+}
+
+// Put a db entry for the provided remote segment info
+func (db *Database) Put(ctx context.Context, rmtSegInfo *RemoteSegmentInfo) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	// entry exists already?
+	updateResp, err := db.Get(ctx, rmtSegInfo)
+	if err != nil {
+		return db.Create(ctx, rmtSegInfo)
+	}
+	return db.Update(ctx, updateResp)
+}
+
+// Create a db entry for the provided remote segment info
+func (db *Database) Create(ctx context.Context, info *RemoteSegmentInfo) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	defer db.locked()()
+
+	_, err = db.DB.Create_Irreparabledb(
+		ctx,
+		dbx.Irreparabledb_Segmentpath(info.SegmentPath),
+		dbx.Irreparabledb_Segmentdetail(info.SegmentDetail),
+		dbx.Irreparabledb_PiecesLostCount(info.LostPiecesCount),
+		dbx.Irreparabledb_SegDamagedUnixSec(info.RepairUnixSec),
+		dbx.Irreparabledb_SegCreatedAt(time.Unix(info.RepairUnixSec, 0)),
+		dbx.Irreparabledb_RepairAttemptCount(info.RepairAttemptCount),
+	)
+	if err != nil {
+		return status.Errorf(codes.Internal, err.Error())
+	}
+
+	return nil
+}
+
+// Get a irreparable's segment info from the db
+func (db *Database) Get(ctx context.Context, getReq *RemoteSegmentInfo) (resp *RemoteSegmentInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+	defer db.locked()()
+
+	dbSegInfo, err := db.DB.Get_Irreparabledb_By_Segmentpath(ctx, dbx.Irreparabledb_Segmentpath(getReq.SegmentPath))
+	if err != nil {
+		return &RemoteSegmentInfo{}, status.Errorf(codes.Internal, err.Error())
+	}
+
+	rmtseginfo := &RemoteSegmentInfo{
+		SegmentPath:        dbSegInfo.Segmentpath,
+		SegmentDetail:      dbSegInfo.Segmentdetail,
+		LostPiecesCount:    dbSegInfo.PiecesLostCount,
+		RepairUnixSec:      dbSegInfo.SegDamagedUnixSec,
+		RepairAttemptCount: dbSegInfo.RepairAttemptCount,
+	}
+	return rmtseginfo, nil
+}
+
+// Delete a irreparable's segment info from the db
+func (db *Database) Delete(ctx context.Context, delReq *RemoteSegmentInfo) (err error) {
```

Comment:
Only provide arguments that are actually used for the query.

Reply:
done